subsys/net/lib/dhcpv4/dhcpv4_server.c

```diff
@@ -0,0 +1,1421 @@
+/** @file
+ * @brief DHCPv4 server implementation
+ */
+
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/net_ip.h>
+#include <zephyr/net/dhcpv4.h>
+#include <zephyr/net/dhcpv4_server.h>
+#include <zephyr/net/ethernet.h>
+#include <zephyr/net/socket.h>
+#include <zephyr/net/socket_service.h>
+#include <zephyr/sys/byteorder.h>
+
+LOG_MODULE_REGISTER(net_dhcpv4_server, CONFIG_NET_DHCPV4_SERVER_LOG_LEVEL);
+
+#include "dhcpv4.h"
+#include "net_private.h"
+#include "../../l2/ethernet/arp.h"
+
+#define DHCPV4_OPTIONS_MSG_TYPE_SIZE 3
+#define DHCPV4_OPTIONS_IP_LEASE_TIME_SIZE 6
+#define DHCPV4_OPTIONS_SERVER_ID_SIZE 6
+#define DHCPV4_OPTIONS_SUBNET_MASK_SIZE 6
+#define DHCPV4_OPTIONS_CLIENT_ID_MIN_SIZE 2
+
+#define ADDRESS_RESERVED_TIMEOUT K_SECONDS(5)
+
+/* RFC 1497 [17] */
+static const uint8_t magic_cookie[4] = { 0x63, 0x82, 0x53, 0x63 };
+
+struct dhcpv4_server_ctx {
+	struct net_if *iface;
+	int sock;
+	struct k_work_delayable timeout_work;
+	struct dhcpv4_addr_slot addr_pool[CONFIG_NET_DHCPV4_SERVER_ADDR_COUNT];
+	struct in_addr server_addr;
+	struct in_addr netmask;
+};
+
+static struct dhcpv4_server_ctx server_ctx[CONFIG_NET_DHCPV4_SERVER_INSTANCES];
+static struct zsock_pollfd fds[CONFIG_NET_DHCPV4_SERVER_INSTANCES];
+static K_MUTEX_DEFINE(server_lock);
+
+#define DHCPV4_MAX_PARAMETERS_REQUEST_LEN 16
+
+struct dhcpv4_parameter_request_list {
+	uint8_t list[DHCPV4_MAX_PARAMETERS_REQUEST_LEN];
+	uint8_t count;
+};
+
+static void dhcpv4_server_timeout_recalc(struct dhcpv4_server_ctx *ctx)
+{
+	k_timepoint_t next = sys_timepoint_calc(K_FOREVER);
+	k_timeout_t timeout;
+
+	for (int i = 0; i < ARRAY_SIZE(ctx->addr_pool); i++) {
+		struct dhcpv4_addr_slot *slot = &ctx->addr_pool[i];
+
+		if (slot->state == DHCPV4_SERVER_ADDR_RESERVED ||
+		    slot->state == DHCPV4_SERVER_ADDR_ALLOCATED) {
+			if (sys_timepoint_cmp(slot->expiry, next) < 0) {
+				next = slot->expiry;
+			}
+		}
+	}
+
+	timeout = sys_timepoint_timeout(next);
+
+	if (K_TIMEOUT_EQ(timeout, K_FOREVER)) {
+		LOG_DBG("No more addresses, canceling timer");
+		k_work_cancel_delayable(&ctx->timeout_work);
+	} else {
+		k_work_reschedule(&ctx->timeout_work, timeout);
+	}
+}
+
+/* Option parsing. */
+
+static uint8_t *dhcpv4_find_option(uint8_t *data, size_t datalen,
+				   uint8_t *optlen, uint8_t opt_code)
+{
+	uint8_t *opt = NULL;
+
+	while (datalen > 0) {
+		uint8_t code;
+		uint8_t len;
+
+		code = *data;
+
+		/* Two special cases (fixed sized options) */
+		if (code == 0) {
+			data++;
+			datalen--;
+			continue;
+		}
+
+		if (code == DHCPV4_OPTIONS_END) {
+			break;
+		}
+
+		/* Length field should now follow. */
+		if (datalen < 2) {
+			break;
+		}
+
+		len = *(data + 1);
+
+		if (datalen < len + 2) {
+			break;
+		}
+
+		if (code == opt_code) {
+			/* Found the option. */
+			opt = data;
```

Comment:
Nit: Already set `opt` to the correct offset?

```suggestion
			opt = data + 2;
```

Reply:
Right, done.